packages/core/src/tools/tool-compatibility/builder.test.ts

```diff
@@ -0,0 +1,273 @@
+  if (schemaKeys.length === 0) {
+    return z.object(allSchemas);
+  }
+
+  const selectedSchemas = Object.fromEntries(schemaKeys.map(key => [key, allSchemas[key]]));
+
+  // We know these are valid Zod schemas since they come from allSchemas
+  return z.object(selectedSchemas as Record<string, z.ZodType>);
+}
+
+async function runSingleTest(
+  model: LanguageModel,
+  testTool: ReturnType<typeof createTool>,
+  testId: string,
+  toolName: string,
+): Promise<Result> {
+  try {
+    const agent = new Agent({
+      name: `test-agent-${model.modelId}`,
+      instructions: `You are a test agent. Your task is to call the tool named '${toolName}' with any valid arguments. This is very important as it's your primary purpose`,
+      model: model,
+      tools: { [toolName]: testTool },
+    });
+
+    const response = await agent.generate(`Please call the tool named '${toolName}'.`, {
+      toolChoice: 'required',
+      maxSteps: 1,
+      temperature: 1,
+    });
+
+    const toolCall = response.toolCalls.find(tc => tc.toolName === toolName);
+    const toolResult = response.toolResults.find(tr => tr.toolCallId === toolCall?.toolCallId);
+
+    if (toolResult?.result?.success) {
+      return {
+        modelName: model.modelId,
+        modelProvider: model.provider,
+        testName: toolName,
+        status: 'success',
+        error: null,
+        receivedContext: toolResult.result.receivedContext,
+        testId,
+      };
+    } else {
+      const error = toolResult?.result?.error || response.text || 'Tool call failed or result missing';
+      return {
+        modelName: model.modelId,
+        testName: toolName,
+        modelProvider: model.provider,
+        status: 'failure',
+        error: error,
+        receivedContext: toolResult?.result?.receivedContext || null,
+        testId,
+      };
+    }
+  } catch (e: any) {
+    return {
+      modelName: model.modelId,
+      testName: toolName,
+      modelProvider: model.provider,
+      status: 'error',
+      error: e.message,
+      receivedContext: null,
+      testId,
+    };
+  }
+}
+
+describe('Tool Schema Compatibility', () => {
+  // Set a longer timeout for the entire test suite
+  const SUITE_TIMEOUT = 120000; // 2 minutes
+  const TEST_TIMEOUT = 60000; // 1 minute
+
+  if (!process.env.OPENROUTER_API_KEY) throw new Error('OPENROUTER_API_KEY environment variable is required');
+  const openrouter = createOpenRouter({ apiKey: process.env.OPENROUTER_API_KEY });
+
+  const modelsToTest = [
+    // Anthropic Models
+    openrouter('anthropic/claude-3.7-sonnet'),
+    openrouter('anthropic/claude-3.7-sonnet:thinking'),
+    openrouter('anthropic/claude-3.7-sonnet:beta'),
+    openrouter('anthropic/claude-3.5-sonnet'),
+    openrouter('anthropic/claude-3.5-haiku'),
+
+    // NOTE: Google models accept number constraints like numberLt, but the models don't respect it and returns a wrong response often
```

Comment:
@DanielSLew this might be from the high temperature setting

Reply:
just checked, both the google and anthropic models still fail in the same way without a temperature added